CMSgov/easi-app
# [EASI-3360] GraphQL schema for TRB Admin Note category-specific data

# EASI-3360

## Changes and Description

- Update the GraphQL schema to include category-specific data on TRB admin notes (i.e. notes in the Initial Request Form category can select which sections the note applies to)
  - For queries:
    - Each category has a type for the data that's specific to that category (including the General Request and Consult Session categories, which _don't_ have any category-specific data).
    - The `TRBAdminNote` type has a `categorySpecificData` field whose type is a [GraphQL union](https://graphql.org/learn/schema/#union-types) of the category-specific types.
  - For mutations:
    - Each category has its own mutation and associated input type, with the types having non-nullable fields for required data.
      - The resolvers for the new mutations have placeholder implementations that call the existing `CreateTRBAdminNote()` function, so the frontend can more easily code against them in EASI-3361.
    - The existing `createTRBAdminNote` mutation is kept around to avoid breaking existing code; it will be removed in EASI-3458.
    - The `updateTRBAdminNote` mutation and resolver don't currently account for category-specific data. That mutation isn't currently being used, but might be in the future (see [this Slack conversation](https://cmsgov.slack.com/archives/CNU2B59UH/p1697562260732339)); if we need to update it to allow editing category-specific data, we can change the schema at that time.
- Provide mock data for the `categorySpecificData` field's resolver on the backend, including category-specific data, for the frontend to use as an example to code against.
- Update the Postman collection to include an example of querying the union on the `TRBAdminNote` type, as well as saving requests to call `createTRBAdminNote` and `updateTRBAdminNote`.

## How to test this change

There's not much to test, since the bulk of this PR is GraphQL schema changes that haven't been implemented. However, we can test that the existing `createTRBAdminNote` calls still work, and that the placeholder data is returned.
1. Spin up the app, including the frontend.
2. Run `scripts/dev db:seed`.
3. In the UI, go to the TRB admin view, open "Case 7 - Advice letter in review".
4. View the advice letter and add a note (of any category). The note should be created without any issues, and should be able to be viewed without errors.
5. Open Postman, set the `trbRequestID` collection variable's current value to the ID of the "Case 7" request, and save that value.
6. Run the "Get Admin Notes, with category-specific fields" request (under TRB -> TRB Admin Notes). This should return the placeholder data from `schema.resolvers.go` for the note's category.

## Questions
1. As mentioned above, I'm not sure whether to have types for all five categories, including the categories that don't currently have category-specific data, and include them in the `TRBAdminNoteCategorySpecificData` union. Including all five makes it clear that it covers all categories, but having the extra types that don't have any useful data is annoying, especially because gqlgen doesn't support empty types (hence the `placeholderField` fields). It doesn't add any major burden either way to the backend; @adamodd, @aterstriep, @patrickseguraoddball, what do you think?
2. I've nested common fields used by all categories of advice notes in the `CreateTRBAdminNoteCommonFields` input type to avoid repetition and redundancy - @aterstriep, @adamodd, @patrickseguraoddball, does this sort of nesting work fine with frontend code that calls the mutation from a form?

## Notes
* Backend implementation notes (for EASI-3362):
  * We'll need to add fields to the `TRBAdminNote` Go model and SQL table for the data that can go right on the intake (such as the various "appliesTo*" checkboxes/booleans). The relations with TRB documents and recommendations will need to be handled by many-to-many link tables in SQL, with small Go models representing each link, similar to [`/pkg/models/trb_request_lcid.go`](https://github.com/CMSgov/easi-app/blob/main/pkg/models/trb_request_lcid.go), which links TRB requests and system intakes.
  * The SQL columns added to `trb_admin_notes` will be nullable, by necessity; we'll see in EASI-3362 what constraints can be enforced in SQL.
  * Existing admin notes can just have all `appliesTo*` checkboxes set to false (when applicable), with no links to documents/recommendations created for them.
  * More details are in [this Jira comment](https://jiraent.cms.gov/browse/EASI-3360?focusedId=2356345&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2356345) on EASI-3360.

## Decisions after PR was initially posted for review
- Dev meeting on 10/17/23 decided to use multiple mutations, one per category, for creating admin notes.
  -  GraphQL doesn't currently support union types in input fields - the original approach was to have nullable fields covering all possible categories in `CreateTRBAdminNoteInput`, with a single resolver. We went with the approach of multiple mutations for greater type-safety.
- Clay and I decided on 10/17/23 to wait on implementing category-specific functionality for `updateTRBAdminNote` for now, after talking with UX.
